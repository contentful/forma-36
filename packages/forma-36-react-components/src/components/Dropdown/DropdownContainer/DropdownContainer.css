@import 'settings/colors';
@import 'settings/typography';
@import 'settings/transitions';
@import 'settings/dimensions';
@import 'settings/shadows';
@import 'settings/borders';
@import 'settings/z-index';

.DropdownContainer {
  background: var(--color-white);
  box-sizing: border-box;
  border: 0;
  border-radius: var(--border-radius-medium);
  box-shadow: var(--box-shadow-default);
  list-style: none;
  padding: 0;
  /*
   * @todo: Remove this in next breaking change, it's overriden by Popper,
   * who uses absolute positioning
   */
  position: fixed;
  margin: 0;

  z-index: var(--z-index-dropdown);

  &:focus {
    outline: none;
    box-shadow: var(--glow-primary), var(--box-shadow-default);
  }
}

.DropdownContainer > div {
  width: 100%;
}
