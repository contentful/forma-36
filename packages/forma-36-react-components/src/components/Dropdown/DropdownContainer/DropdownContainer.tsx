import React, { Component } from 'react';
import cn from 'classnames';
import ReactDOM from 'react-dom';
import InViewport from '../../InViewport';
import {
  positionType,
  AnchorDimensionsAndPositonType,
} from '../Dropdown/Dropdown';

const styles = require('./DropdownContainer.css');

export type DropdownContainerProps = {
  onClose?: Function;
  dropdownAnchor?: HTMLElement | null;
  className?: string;
  children?: React.ReactNode;
  testId?: string;
  openSubmenu?: (value: boolean) => void;
  anchorDimensionsAndPositon?: AnchorDimensionsAndPositonType;
  position: positionType;
  getRef?: (ref: HTMLElement | null) => void;
  submenu?: boolean;
  width?: number;
  isAutoalignmentEnabled?: boolean;
} & typeof defaultProps;

export interface DropdownState {
  dropdownDimensions: {
    width: number;
    height: number;
  };
  position: positionType;
}

const defaultProps = {
  testId: 'cf-ui-dropdown-portal',
  position: 'bottom-left',
  submenu: false,
  isAutoalignmentEnabled: true,
  getRef: () => {},
};

class DropdownContainer extends Component<
  DropdownContainerProps,
  DropdownState
> {
  static defaultProps = defaultProps;

  portalTarget = document.createElement('div');
  dropdown: HTMLElement | null = null;
  lastOverflowAt: string | null = null;

  state = {
    position: this.props.position,
    dropdownDimensions: {
      width: 0,
      height: 0,
    },
  };

  componentDidMount() {
    document.body.appendChild(this.portalTarget);
    if (this.dropdown) {
      const dropdownRect = this.dropdown.getBoundingClientRect();
      this.setState({
        dropdownDimensions: {
          width: dropdownRect.width,
          height: dropdownRect.height,
        },
      });
    }
    document.addEventListener('mousedown', this.trackOutsideClick, true);
    this.props.getRef(this.dropdown);
  }

  componentWillUnmount() {
    document.body.removeChild(this.portalTarget);
    document.removeEventListener('mousedown', this.trackOutsideClick, true);
  }

  trackOutsideClick = (e: MouseEvent) => {
    if (
      this.dropdown &&
      !this.dropdown.contains(e.target as Node) &&
      (this.props.dropdownAnchor &&
        !this.props.dropdownAnchor.contains(e.target as Node))
    ) {
      if (this.props.onClose) {
        this.props.onClose();
      }
    }
  };

  handleOverflow = (overflowAt: string) => {
    if (!this.props.isAutoalignmentEnabled) {
      return;
    }
    if (overflowAt === this.lastOverflowAt) {
      return;
    }

    const resolutions = {
      right: {
        'bottom-left': 'bottom-right',
        'top-left': 'top-right',
        right: 'left',
      },
      left: {
        'bottom-right': 'bottom-left',
        'top-right': 'top-left',
        left: 'right',
      },
      top: {
        'top-left': 'bottom-left',
        'top-right': 'bottom-right',
      },
      bottom: {
        'bottom-left': 'top-left',
        'bottom-right': 'top-right',
      },
    };
    const currentPosition = this.state.position;
    const resolution = resolutions[overflowAt][currentPosition];
    if (resolution) {
      this.setState(
        {
          position: resolution,
        },
        () => {
          this.lastOverflowAt = overflowAt;
        },
      );
    }
  };

  calculatePosition = () => {
    const { anchorDimensionsAndPositon } = this.props;
    const { dropdownDimensions, position } = this.state;

    if (!anchorDimensionsAndPositon || !dropdownDimensions) {
      return false;
    }

    switch (position) {
      case 'bottom-left':
        return {
          top:
            anchorDimensionsAndPositon.top + anchorDimensionsAndPositon.height,
          left: anchorDimensionsAndPositon.left,
        };
      case 'top-left':
        return {
          bottom: window.innerHeight - anchorDimensionsAndPositon.top,
          left: anchorDimensionsAndPositon.left,
        };
      case 'bottom-right':
        return {
          top:
            anchorDimensionsAndPositon.top + anchorDimensionsAndPositon.height,
          left:
            anchorDimensionsAndPositon.left -
            (dropdownDimensions.width - anchorDimensionsAndPositon.width),
        };
      case 'top-right':
        return {
          bottom: window.innerHeight - anchorDimensionsAndPositon.top,
          left:
            anchorDimensionsAndPositon.left -
            (dropdownDimensions.width - anchorDimensionsAndPositon.width),
        };
    }
  };

  getSubmenuClassNames = () =>
    cn(
      styles['DropdownContainer__submenu'],
      styles[`DropdownContainer__container-position--${this.state.position}`],
    );

  render() {
    const { submenu, className, width } = this.props;

    const classNames = cn(
      className,
      styles['DropdownContainer'],
      submenu ? this.getSubmenuClassNames() : '',
    );

    const dropdown = (
      <div
        ref={ref => {
          this.dropdown = ref;
        }}
        style={{
          ...(width ? { width: `${width}px` } : {}),
          ...(!submenu && this.calculatePosition()),
        }}
        className={classNames}
        onMouseEnter={() => {
          if (this.props.openSubmenu) {
            this.props.openSubmenu(true);
          }
        }}
        onFocus={() => {
          if (this.props.openSubmenu) {
            this.props.openSubmenu(true);
          }
        }}
        onMouseLeave={() => {
          if (this.props.openSubmenu) {
            this.props.openSubmenu(false);
          }
        }}
      >
        <InViewport
          onOverflowLeft={() => this.handleOverflow('left')}
          onOverflowRight={() => this.handleOverflow('right')}
          onOverflowTop={() => this.handleOverflow('top')}
          onOverflowBottom={() => this.handleOverflow('bottom')}
        >
          {this.props.children}
        </InViewport>
      </div>
    );

    return submenu
      ? dropdown
      : ReactDOM.createPortal(dropdown, this.portalTarget);
  }
}

export default DropdownContainer;
