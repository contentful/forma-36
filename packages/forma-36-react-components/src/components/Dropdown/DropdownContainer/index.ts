export * from './DropdownContainer';
export { default } from './DropdownContainer';
