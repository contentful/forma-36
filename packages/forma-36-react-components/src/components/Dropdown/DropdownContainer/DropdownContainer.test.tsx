import React from 'react';
import { shallow } from 'enzyme';
import DropdownContainer from './DropdownContainer';

it('renders the component', () => {
  const output = shallow(
    <DropdownContainer isOpen>DropdownContainer</DropdownContainer>,
  );
  expect(output).toMatchSnapshot();
});

it('renders the component with an additional class name', () => {
  const output = shallow(
    <DropdownContainer className="extraClassName" isOpen>
      DropdownContainer
    </DropdownContainer>,
  );
  expect(output).toMatchSnapshot();
});

it('renders the component as a submenu', () => {
  const output = shallow(
    <DropdownContainer className="extraClassName" isOpen submenu>
      DropdownContainer
    </DropdownContainer>,
  );
  expect(output).toMatchSnapshot();
});

it('has no a11y issues', async () => {
  const output = shallow(
    <DropdownContainer isOpen>DropdownContainer</DropdownContainer>,
  );
  expect(output).toMatchSnapshot();
});
