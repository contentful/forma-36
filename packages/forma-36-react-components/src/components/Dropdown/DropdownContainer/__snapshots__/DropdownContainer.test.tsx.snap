// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`has no a11y issues 1`] = `
<Portal
  containerInfo={<div />}
>
  <div
    anchorDimensionsAndPositon={
      Object {
        "height": 0,
        "left": 0,
        "top": 0,
        "width": 100,
      }
    }
    className="DropdownContainer"
    getRef={[Function]}
    onFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    position="bottom-left"
    style={
      Object {
        "left": 0,
        "top": 0,
      }
    }
    testId="cf-ui-dropdown-portal"
  >
    <InViewport
      offset={0}
      onOverflowBottom={[Function]}
      onOverflowLeft={[Function]}
      onOverflowRight={[Function]}
      onOverflowTop={[Function]}
      testId="cf-ui-in-viewport"
    >
      DropdownContainer
    </InViewport>
  </div>
</Portal>
`;

exports[`renders the component 1`] = `
<Portal
  containerInfo={<div />}
>
  <div
    anchorDimensionsAndPositon={
      Object {
        "height": 0,
        "left": 0,
        "top": 0,
        "width": 100,
      }
    }
    className="DropdownContainer"
    getRef={[Function]}
    onFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    position="bottom-left"
    style={
      Object {
        "left": 0,
        "top": 0,
      }
    }
    testId="cf-ui-dropdown-portal"
  >
    <InViewport
      offset={0}
      onOverflowBottom={[Function]}
      onOverflowLeft={[Function]}
      onOverflowRight={[Function]}
      onOverflowTop={[Function]}
      testId="cf-ui-in-viewport"
    >
      DropdownContainer
    </InViewport>
  </div>
</Portal>
`;

exports[`renders the component as a submenu 1`] = `
<div
  anchorDimensionsAndPositon={
    Object {
      "height": 0,
      "left": 0,
      "top": 0,
      "width": 100,
    }
  }
  className="extraClassName DropdownContainer DropdownContainer__submenu DropdownContainer__container-position--bottom-left"
  getRef={[Function]}
  onFocus={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  position="bottom-left"
  style={Object {}}
  testId="cf-ui-dropdown-portal"
>
  <InViewport
    offset={0}
    onOverflowBottom={[Function]}
    onOverflowLeft={[Function]}
    onOverflowRight={[Function]}
    onOverflowTop={[Function]}
    testId="cf-ui-in-viewport"
  >
    DropdownContainer
  </InViewport>
</div>
`;

exports[`renders the component with an additional class name 1`] = `
<Portal
  containerInfo={<div />}
>
  <div
    anchorDimensionsAndPositon={
      Object {
        "height": 0,
        "left": 0,
        "top": 0,
        "width": 100,
      }
    }
    className="extraClassName DropdownContainer"
    getRef={[Function]}
    onFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    position="bottom-left"
    style={
      Object {
        "left": 0,
        "top": 0,
      }
    }
    testId="cf-ui-dropdown-portal"
  >
    <InViewport
      offset={0}
      onOverflowBottom={[Function]}
      onOverflowLeft={[Function]}
      onOverflowRight={[Function]}
      onOverflowTop={[Function]}
      testId="cf-ui-in-viewport"
    >
      DropdownContainer
    </InViewport>
  </div>
</Portal>
`;
